homeassistant/components/econet/water_heater.py

```diff
@@ -1,153 +1,82 @@
-
-    econet = PyEcoNet(username, password)
-    water_heaters = econet.get_water_heaters()
-    hass_water_heaters = [
-        EcoNetWaterHeater(water_heater) for water_heater in water_heaters
-    ]
-    add_entities(hass_water_heaters)
-    hass.data[ECONET_DATA]["water_heaters"].extend(hass_water_heaters)
-
-    def service_handle(service):
-        """Handle the service calls."""
-        entity_ids = service.data.get("entity_id")
-        all_heaters = hass.data[ECONET_DATA]["water_heaters"]
-        _heaters = [
-            x for x in all_heaters if not entity_ids or x.entity_id in entity_ids
-        ]
-
-        for _water_heater in _heaters:
-            if service.service == SERVICE_ADD_VACATION:
-                start = service.data.get(ATTR_START_DATE)
-                end = service.data.get(ATTR_END_DATE)
-                _water_heater.add_vacation(start, end)
-            if service.service == SERVICE_DELETE_VACATION:
-                for vacation in _water_heater.water_heater.vacations:
-                    vacation.delete()
-
-            _water_heater.schedule_update_ha_state(True)
-
-    hass.services.register(
-        DOMAIN, SERVICE_ADD_VACATION, service_handle, schema=ADD_VACATION_SCHEMA
-    )
 
-    hass.services.register(
-        DOMAIN, SERVICE_DELETE_VACATION, service_handle, schema=DELETE_VACATION_SCHEMA
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up EcoNet water heater based on a config entry."""
+    equipment = hass.data[DOMAIN][EQUIPMENT][entry.entry_id]
+    async_add_entities(
+        [
+            EcoNetWaterHeater(water_heater)
+            for water_heater in equipment[EquipmentType.WATER_HEATER]
+        ],
     )
 
 
-class EcoNetWaterHeater(WaterHeaterEntity):
-    """Representation of an EcoNet water heater."""
+class EcoNetWaterHeater(EcoNetEntity, WaterHeaterEntity):
+    """Define a Econet water heater."""
 
     def __init__(self, water_heater):
-        """Initialize the water heater."""
+        """Initialize."""
+        super().__init__(water_heater)
+        self._running = water_heater.running
+        self._poll = True
         self.water_heater = water_heater
-        self.supported_modes = self.water_heater.supported_modes
         self.econet_state_to_ha = {}
         self.ha_state_to_econet = {}
-        for mode in ECONET_STATE_TO_HA:
-            if mode in self.supported_modes:
-                self.econet_state_to_ha[mode] = ECONET_STATE_TO_HA.get(mode)
-        for key, value in self.econet_state_to_ha.items():
-            self.ha_state_to_econet[value] = key
-        for mode in self.supported_modes:
-            if mode not in ECONET_STATE_TO_HA:
-                error = f"Invalid operation mode mapping. {mode} doesn't map. Please report this."
-                _LOGGER.error(error)
 
-    @property
-    def name(self):
-        """Return the device name."""
-        return self.water_heater.name
+    def on_update_received(self):
```

Comment:
Can we change this to run in async context, by decorating it with `@callback`? Are the `water_heater` access async safe?

Reply:
All the properties are just values in a dictionary in pyeconet.

Reply:
Good! Then we can make it async.

Reply:
```suggestion
    @callback
    def on_update_received(self):
```